uroute/url.py

```diff
@@ -67,31 +67,37 @@ def clean_url(self, url):
             if not re.match(provider['urlPattern'], url, re.IGNORECASE):
                 continue
 
+            # If any exceptions are matched, this provider is skipped
             if any(
                 re.match(exc, url, re.IGNORECASE)
-                for exc in provider['exceptions']
+                for exc in provider.get('exceptions', [])
             ):
                 continue
 
-            for redir in provider['redirections']:
+            # If redirect found, recurse on target (only once)
+            for redir in provider.get('redirections', []):
                 match = re.match(redir, url, re.IGNORECASE)
                 try:
                     if match and match.group(1):
-                        return unquote(match.group(1))
+                        if loop:
+                            return self.clean_url(unquote(match.group(1)),
+                                                  False)
+                        else:
+                            url = unquote(match.group(1))
```

Comment:
I would structure this change a little bit differently:

```suggestion
                        url = unquote(match.group(1))
                        if loop:
                            return self.clean_url(url, loop=False)
                        return url
```

There's a few things to note here:

1. This version deduplicates the `unquote(...)` call ([DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)).
1. It's common (and clearer IMHO) to add keyword argument names (`loop=`) in function calls.
1. The else-branch of your change doesn't `return` the matched value, as the original code does. If that was intended, would you mind explaining why that change is necessary?

Reply:
Great notes! Regarding point 3 that was unintentional, I must have introduced that error when cleaning this up for the pull. Oops!